drivers/MinimalSerial.h

```diff
@@ -0,0 +1,124 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2019 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_MINIMAL_SERIAL_H
+#define MBED_MINIMAL_SERIAL_H
+
+#include "platform/platform.h"
+
+#if DEVICE_SERIAL || defined(DOXYGEN_ONLY)
+#include "hal/serial_api.h"
+#include "platform/NonCopyable.h"
+#include "PinNames.h"
+
+namespace mbed {
+
+/**
+ * \defgroup drivers_MinimalSerial MinimalSerial class
+ * \ingroup drivers-public-api-uart
+ * @{
+ */
+
+/** A abstract class for serial port implementations
+ */
+class MinimalSerial : private NonCopyable<MinimalSerial> {
+
+public:
+
+    enum Flow {
+        Disabled = 0,
+        RTS,
+        CTS,
+        RTSCTS
+    };
+
+#if DEVICE_SERIAL_FC
+    /** Set the flow control type on the serial port
+     *
+     *  @param type the flow control type (Disabled, RTS, CTS, RTSCTS)
+     *  @param flow1 the first flow control pin (RTS for RTS or RTSCTS, CTS for CTS)
+     *  @param flow2 the second flow control pin (CTS for RTSCTS)
+     */
+    void set_flow_control(Flow type, PinName flow1 = NC, PinName flow2 = NC);
+#endif // DEVICE_SERIAL_FC
+
+    /** Set the baud rate of the serial port
+     *
+     *  @param baudrate The baudrate of the serial port.
+     */
+    void baud(int baudrate);
+
+    /** Write a char to the serial port
+     *
+     * @param c The char to write
+     *
+     * @returns The written char
+     */
+    int putc(int c);
```

Comment:
If `MinimalSerial` is to be the base class of `MinimalConsole` and `RawSerial` (`UnbufferedSerial`) then it must have `putc` and `getc` to avoid repetition in those two classes.
`SerialBase` does not currently implement them because oinly `RawSerial` implements them.

Reply:
MinimalConsole should use _base_putc. Also UARTSerial/BufferedSerial does not have putc/getc. Only UnbufferedSerial had it and Kevin suggested that they shoudl be removed which I agree, see https://github.com/ARMmbed/mbed-os/pull/11634/files/06d69f778b8dc87575ead010049e03478b0d86b7#r334879444

Reply:
Surely the `lock/unlock` is needed to do putc/getc in `MinimalConsole`. `_base_putc` assumes mutex lock is already held so `MinimalConsole` must have somehow held a mutex before calling `_base_putc` thus `putc`.

Reply:
You have defined lock and unlock (with empty definition). No issue with that.

Reply:
They are virtual functions to be replaced in their subclasses.